File: templating/tag_contributions/src/main/java/com/vmware/jsonteng/ext/tags/net/Ipv4HostGatewayTag.java
package com.vmware.jsonteng.ext.tags.net;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.TagResolver;
import com.vmware.jsonteng.TemplateEngineException;
import com.vmware.jsonteng.tags.TagBase;

public class Ipv4HostGatewayTag extends TagBase {
    public static final String name = "ipv4-host-gateway";

    private final ElementResolver elementResolver;

    public Ipv4HostGatewayTag(TagResolver TagResolver) {
        super(TagResolver);
        elementResolver = TagResolver.getElementResolver();
    }

    @Override
    public Object process(List<?> tagTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        int tokenCount = tagTokens.size();
        if (tokenCount != 1) {
            throw new TemplateEngineException(
                    String.format("Tag \"%s\" requires 1 parameter. Parameters given %s", name, tagTokens));
        }
        String resolvedNetwork = (String) this.elementResolver.resolve(
                tagTokens.get(0), bindingDataList);
        String[] parts = resolvedNetwork.split("/");
        Inet4Address networkAddress;
        try {
            networkAddress = (Inet4Address) InetAddress.getByName(parts[0]);
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unable to get the network address " + parts[0], e);
        }
        ByteBuffer fourBytes = ByteBuffer.wrap(new byte[4]);
        fourBytes.put(networkAddress.getAddress());
        fourBytes.rewind();
        int network = fourBytes.getInt();
        int gatewayInt = network + 1;
        fourBytes.rewind();
        fourBytes.putInt(gatewayInt);
        Inet4Address gateway;
        try {
            gateway = (Inet4Address) InetAddress.getByAddress(fourBytes.array());
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unabled to get ip address " + fourBytes, e);
        }
        return gateway.getHostAddress();
    }
}

File: templating/tag_contributions/src/main/go/src/jsonteng_contribs/tags/ipv4_host_gateway_tag.go
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package tags

import (
	"container/list"
	"fmt"
	"jsonteng/errors"
	intfs "jsonteng/interfaces"
	"net"
)

type ipv4HostGatewayTag struct {
	intfs.TagBase
	elementResolver intfs.ElementResolver
}

func (tag *ipv4HostGatewayTag) init(tagResolver intfs.TagResolver) {
	tag.Name = "ipv4-host-gateway"
	tag.TagResolver = tagResolver
	tag.elementResolver = tagResolver.GetElementResolver()
}

func (tag *ipv4HostGatewayTag) Process(tagTokens []interface{}, bindingDataList *list.List) (interface{}, error) {
	tokenCount := len(tagTokens)
	if tokenCount != 1 {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Tag \"%s\" requires 1 parameter. Parameters given %v", tag.Name, tagTokens)
		return nil, &errl
	}
	resolvedNetworkObj, err := tag.elementResolver.ResolveElement(tagTokens[0], bindingDataList)
	if err != nil {
		return nil, err
	}
	resolvedNetwork, ok := resolvedNetworkObj.(string)
	if !ok {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Invalid network %v", resolvedNetworkObj)
		return nil, &errl
	}
	_, ipv4Net, err := net.ParseCIDR(resolvedNetwork)
	if err != nil {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Invalid network format %v", resolvedNetwork)
		return nil, &errl
	}
	network := ipv4Net.IP
	return net.IPv4(network[0], network[1], network[2], network[3]+1).String(), nil
}
